### DISCUSSION TITLE
Is there a way to open the same db in both readonly mode and write mode

### DISCUSSION THREAD

**Discussion by quangson91** [tekartik/sqflite#980] [Q&A] [answered]
I have some usecase that need open db in two modes.
For example: 
- The `trust` query -> can be execute in writable connect.
- The `un trust` query -> My purpose is that user type query for read data. BUT prevent they edit/update anything. So I want to use db readonly in this case.

Eg:

```
final dbWriable = await openDatabase('data.db');
final dbReadOnly = await openReadOnlyDatabase('data.db');
```

But it seems the dbReadonly still able to insert/update/delete data.

---------

So is there a way to do that ?
Or am I missing some config.

Thx!

**Comment by alextekartik** [comment, marked as answer]
By default, openDatabase and openReadOnlyDatabase uses the flag `singleInstance: true` which means "The first to open wins" and any other call to openDatabase will get the same instance (i.e. in your case not read-only). One solution is to set this flag to false. This is not something many people used as it means that you might encounter lock issue (which means sometimes retrying an operation) when the 2 instances access the database at the same time.

What you can try (so unfortunately here you cannot use openReadOnlyDatabase as it does not have this options):

```dart
// Open in write mode
var dbWritable = await openDatabase(path, readOnly: false, singleInstance: false);
// Open in read mode
var dbReadOnly = await openDatabase(path, readOnly: true, singleInstance: false);
```

**Reply by quangson91** [reply]
It worked for me.

May be an extra question.
You said:

> This is not something many people used as it means that you might encounter lock issue (which means sometimes retrying an operation) when the 2 instances access the database at the same time.


So, is there a way that check if query is `writable query` which mean, that query may change the database.

Thank you so much for the fast reply & support! (bow)

**Comment by alextekartik** [comment]
> So, is there a way that check if query is writable query which mean, that query may change the database.

Parsing SQL could be very complex. For simplicity, I would assume that any statement not starting with `SELECT` could potentially change the database

**Reply by quangson91** [reply]
Thank you for the suggestion!
I think I will choose solution open db read-only mode and using WAL mode.

---

Thank you for build the so good library!
